### ISSUE TITLE
Error when querying with ISO output schema

### ISSUE THREAD

**Issue by ludona7** [geopython/pycsw#576]
# Description
When using `GetRecordByID` request with `outputSchema` parameter set to `http://www.isotc211.org/2005/gmd` the server responds with a 500 error. When the `outputSchema` is not set, it returns the record in default schema with no problems.

# Environment

- operating system: CentOS 7.5.1804
- Python version: 3.4.9
- pycsw version: 2.2.0
- source/distribution
  - [ ] git clone
  - [ ] DebianGIS/UbuntuGIS
  - [x] PyPI
  - [ ] zip/tar.gz
  - [ ] other (please specify):
- web server
  - [ ] Apache/mod_wsgi
  - [ ] CGI
  - [x] other (please specify): Gunicorn

# Steps to Reproduce

Example request:
```
GET /?service=CSW&version=2.0.2&request=GetRecordById&elementsetname=full&outputSchema=http:%2F%2Fwww.isotc211.org%2F2005%2Fgmd&id=12345
```

# Additional Information

Inspecting the logs reveals a python TypeError:
```
File "/venvs/pycsw_venv/lib/python3.4/site-packages/pycsw/plugins/profiles/apiso/apiso.py", line 389, in write_record
    if caps is None and xml_blob is not None and xml_blob.startswith(b'<gmd:MD_Metadata'):
TypeError: startswith first arg must be str or a tuple of str, not bytes
```
To get around this I've made adjustments in our running instance of `pycsw/plugins/profiles/apiso/apiso.py`.

Current master version (lines 388-394):
```python
xml_blob = util.getqattr(result, self.context.md_core_model['mappings']['pycsw:XML'])
if caps is None and xml_blob is not None and xml_blob.startswith(b'<gmd:MD_Metadata'):
    is_iso_anyway = True

if (esn == 'full' and (typename == 'gmd:MD_Metadata' or is_iso_anyway)):
    # dump record as is and exit
    return etree.fromstring(xml_blob, self.context.parser)
```

Adjusted version:
```python
xml_blob = util.getqattr(result, self.context.md_core_model['mappings']['pycsw:XML'])
xml_blob_bytes_object = bytes.fromhex(xml_blob[2:]) # Convert long hex value to bytes object (trim \x from start of string so that fromhex works)
xml_blob_decoded = xml_blob_bytes_object.decode('utf-8') # Then decode the bytes to unicode string

if caps is None and xml_blob is not None and xml_blob_decoded.startswith('<gmd:MD_Metadata'): # Removed `b` prefix, as it is checking a str. Also check on new `xml_blob_decoded` variable instead
    is_iso_anyway = True

if (esn == 'full' and (typename == 'gmd:MD_Metadata' or is_iso_anyway)):
    # dump record as is and exit
    return etree.fromstring(xml_blob_decoded, self.context.parser) # use new `xml_blob_decoded` variable that stores real unicode string value of the XML.
```

I get the real XML markup (not hex as per database value) and store in `xml_blob_decoded` because `fromstring` expects to parse real XML markup. I'm unsure if this is the correct way to patch this, but it now allows me to search for a record with ISO output schema at least as a temporary workaround.

**Comment by benjwadams** [comment]
I'm running off of the PyCSW 2.2.0 Docker container on IOOS Catalog and also tracked down to this issue.  Wasn't initially aware this issue existed, but I came to a similar conclusion running against Python 3.  This is only occurring with the apiso profile output.

**Comment by tomkralidis** [comment]
@ludona7 thanks for the report. I tried testing this along with the PR #580 from @juozasg, in order to fix the Travis issues.  I ran across either the fix working locally failing Travis or vice-versa.

Please see #596 which attempts to cover both bytes and string cases accordingly.  This works for me locally and Travis passes (note that you will 2.7 related errors on Travis that are unrelated/based on libxml2/lxml, and that we don't require 2.7 tests to pass anymore).

If this fix works for everyone I can merge as we are closing in on 2.4 release.

**Comment by lumonald** [comment]
Hi @tomkralidis, I had a bit of trouble getting mine to work with the https://github.com/geopython/pycsw/pull/596 fix incorporated. This is the error I see -

```
  File "/venvs/pycsw_venv/lib/python3.4/site-packages/pycsw/plugins/profiles/apiso/apiso.py", line 402, in write_record
    return etree.fromstring(xml_blob, self.context.parser)
  File "src/lxml/lxml.etree.pyx", line 3213, in lxml.etree.fromstring (src/lxml/lxml.etree.c:77721)
  File "src/lxml/parser.pxi", line 1830, in lxml.etree._parseMemoryDocument (src/lxml/lxml.etree.c:116658)
  File "src/lxml/parser.pxi", line 1711, in lxml.etree._parseDoc (src/lxml/lxml.etree.c:115204)
  File "src/lxml/parser.pxi", line 1051, in lxml.etree._BaseParser._parseUnicodeDoc (src/lxml/lxml.etree.c:109329)
  File "src/lxml/parser.pxi", line 584, in lxml.etree._ParserContext._handleParseResultDoc (src/lxml/lxml.etree.c:103568)
  File "src/lxml/parser.pxi", line 694, in lxml.etree._handleParseResult (src/lxml/lxml.etree.c:105222)
  File "src/lxml/parser.pxi", line 624, in lxml.etree._raiseParseError (src/lxml/lxml.etree.c:104131)
  File "<string>", line None
lxml.etree.XMLSyntaxError: Start tag expected, '<' not found, line 1, column 1
```

The new `isinstance` conditional fixes the `startswith(iso_string)` problem, but I'm getting the same problem caught further down in the code on this return statement:

`return etree.fromstring(xml_blob, self.context.parser)`

I think this is because `xml_blob` is **always** a hexadecimal string (for me at least), rather than a string of valid XML or a bytes object. This means the `isinstance` conditional was not needed in my case (not sure if this is necessary to keep in for other cases).

If I perform the same decoding operations as per https://github.com/geopython/pycsw/issues/576#issue-373560799 on `xml_blob` it works for me, looks a bit ugly however :-(

My working example:

```python
def write_record(self, result, esn, outputschema, queryables, caps=None):
    ''' Return csw:SearchResults child as lxml.etree.Element '''
    typename = util.getqattr(result, self.context.md_core_model['mappings']['pycsw:Typename'])
    is_iso_anyway = False

    xml_blob = util.getqattr(result, self.context.md_core_model['mappings']['pycsw:XML'])

    # convert hexadecimal string value to valid XML string
    xml_blob_decoded = bytes.fromhex(xml_blob[2:]).decode('utf-8') # <----- NEW

    if isinstance(xml_blob, bytes):
        iso_string = b'<gmd:MD_Metadata>'
    else:
        iso_string = '<gmd:MD_Metadata>'

    if caps is None and xml_blob is not None and xml_blob.startswith(iso_string):
        is_iso_anyway = True

    if (esn == 'full' and (typename == 'gmd:MD_Metadata' or is_iso_anyway)):
        # dump (decoded) record as is and exit
        return etree.fromstring(xml_blob_decoded, self.context.parser) # <----- CHANGED
```

**Comment by tomkralidis** [comment]
Thanks @lumonald / @ludona7.  Looks like this causes errors on Travis (https://travis-ci.org/geopython/pycsw/builds/526830979). Looks like we still need to dig.

**Comment by klimeto** [comment]
@tomkralidis 

I deployed the last recent release 2.4.0 and still have the problem to return the ISO encoding of the metadata previously inserted via CSW insert in GMD encoding. 

Sample query:

```url
https://klimeto.com/services/pycsw/240?service=CSW&version=2.0.2&request=GetRecordById&outputFormat=application%2Fxml&outputSchema=http://www.isotc211.org/2005/gmd&elementsetname=full&id=https://data.gov.sk/set/rpi/gmd/17316219/SK_UGKK_ZBGIS_ZBGIS
```

Returns Internal Server Error 

If I remove the outputSchema and use the defaul DC it returns the metadata correctly:
```url
https://klimeto.com/services/pycsw/240?service=CSW&version=2.0.2&request=GetRecordById&outputFormat=application%2Fxml&elementsetname=full&id=https://data.gov.sk/set/rpi/gmd/17316219/SK_UGKK_ZBGIS_ZBGIS
```

Do I still need to apply a manual patch to make this working?

Thanks

**Comment by tomkralidis** [comment]
@klimeto is this still an issue?  Did you apply a manual patch from @ludona7 here?

**Comment by tomkralidis** [comment]
Fixed in #596.
